src/python/library/tritonclient/utils/cuda_shared_memory/__init__.py

```diff
@@ -47,81 +58,33 @@ def from_param(cls, value):
-_ccudashm = cdll.LoadLibrary(_ccudashm_path)
-
-_ccudashm_shared_memory_region_create = _ccudashm.CudaSharedMemoryRegionCreate
-_ccudashm_shared_memory_region_create.restype = c_int
-_ccudashm_shared_memory_region_create.argtypes = [
-    _utf8, c_uint64, c_uint64, POINTER(c_void_p)
-]
-_ccudashm_get_raw_handle = _ccudashm.CudaSharedMemoryGetRawHandle
-_ccudashm_get_raw_handle.restype = c_int
-_ccudashm_get_raw_handle.argtypes = [c_void_p, POINTER(c_char_p)]
-_ccudashm_shared_memory_region_set = _ccudashm.CudaSharedMemoryRegionSet
-_ccudashm_shared_memory_region_set.restype = c_int
-_ccudashm_shared_memory_region_set.argtypes = [
-    c_void_p, c_uint64, c_uint64, c_void_p, c_int
-]
-_cshm_get_shared_memory_handle_info = _ccudashm.GetCudaSharedMemoryHandleInfo
-_cshm_get_shared_memory_handle_info.restype = c_int
-_cshm_get_shared_memory_handle_info.argtypes = [
-    c_void_p,
-    POINTER(c_void_p),
-    POINTER(c_uint64),
-    POINTER(c_uint64),
-    POINTER(c_int)
-]
-
-_cshm_cuda_shared_memory_allocate_and_read_to_host_buffer = _ccudashm.CudaSharedMemoryAllocateAndReadToHostBuffer
-_cshm_cuda_shared_memory_allocate_and_read_to_host_buffer.restype = c_int
-_cshm_cuda_shared_memory_allocate_and_read_to_host_buffer.argtypes = [
-    c_void_p, POINTER(c_char_p)
-]
-_cshm_cuda_shared_memory_release_host_buffer = _ccudashm.CudaSharedMemoryReleaseHostBuffer
-_cshm_cuda_shared_memory_release_host_buffer.restype = c_int
-_cshm_cuda_shared_memory_release_host_buffer.argtypes = [c_char_p]
-
-_ccudashm_shared_memory_region_destroy = _ccudashm.CudaSharedMemoryRegionDestroy
-_ccudashm_shared_memory_region_destroy.restype = c_int
-_ccudashm_shared_memory_region_destroy.argtypes = [c_void_p]
-
-_ccudashm_stream_create = _ccudashm.CudaStreamCreate
-_ccudashm_stream_create.restype = c_int
-_ccudashm_stream_create.argtypes = [POINTER(c_void_p)]
-
-_ccudashm_stream_destroy = _ccudashm.CudaStreamDestroy
-_ccudashm_stream_destroy.restype = c_int
-_ccudashm_stream_destroy.argtypes = [c_void_p]
-
-_ccudashm_stream_synchronize = _ccudashm.CudaStreamSynchronize
-_ccudashm_stream_synchronize.restype = c_int
-_ccudashm_stream_synchronize.argtypes = [c_void_p]
-
 allocated_shm_regions = []
 # Internally managed stream for properly synchronizing on DLPack data,
 # the stream will be created / destroyed according to 'allocated_shm_regions'
 # and be reused throughout the process. May revisit for stream pool if
 # asynchronous write on CUDA shared memory region is requested
-_dlpack_stream = c_void_p()
+_dlpack_stream = None
 
 
-def _raise_if_error(errno):
-    """
-    Raise CudaSharedMemoryException if 'err' is non-success.
-    Otherwise return nothing.
-    """
-    if errno.value != 0:
-        ex = CudaSharedMemoryException(errno)
-        raise ex
-    return
+# Helper function to retrieve internally managed CUDA stream
+def _get_or_create_global_cuda_stream():
+    global _dlpack_stream
+    if _dlpack_stream is None:
+        _dlpack_stream = CudaStream()
```

Comment:
We should probably change global stream to a global pool of streams per device, if we are supporting multi-device

Reply:
Currently the stream is used for data validity, I don't think stream is associated with a particular device, so we don't need to create stream per device (can be more or less than # device), which means we can use only one stream for multiple devices. Having stream pool is beneficial when we want to provide concurrency on reading / writing the CUDA shared memory region, and I think that can be postponed until being requested.

Reply:
I believe CUDA stream is per device, you can't access stream X created on GPU 0 from GPU 1: https://stackoverflow.com/questions/31474784/are-cuda-streams-device-associated-and-how-do-i-get-a-streams-device

Reply:
Thanks for pointing this out, now that you mentioned it, it makes sense as kernel launches with stream must be on some device.

After reading up [this](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#stream-and-event-behavior), seems like memory copies and synchronization is not restricted on the current device and the device that the stream is associated with.

Reply:
Interesting, thank you for linking this! My concern is they say nothing about `cudaStreamSynchronize` (so if we use it, I think it is better to test it to be safe). My second concern, when we are the consumer and pass stream to the provider, I believe they expect stream to be on the same device as specified in the capsule. Since there is no agreement on proper synchronization approach, there may be logic involving proper device - stream pair association. cc @Tabrizian